{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "import msprime\n",
    "from scipy import linalg, stats\n",
    "import utils\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import pysam\n",
    "import pandas as pd\n",
    "from typing import Callable\n",
    "\n",
    "np.random.seed(42)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate data and clean real data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean old data\n",
    "\n",
    "vcf_path = \"raw_data/ALL.chr21.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz\"\n",
    "vcf_file = pysam.VariantFile(vcf_path)\n",
    "\n",
    "records = [record for record in vcf_file.fetch()]\n",
    "\n",
    "\n",
    "# Filter data\n",
    "\n",
    "#16 minutes: 154k observations\n",
    "filtered_by_maf = utils.filter_maf(records)\n",
    "#2 minutes: no change\n",
    "filtered_by_missingness = utils.filter_missingness(filtered_by_maf)\n",
    "#13 seconds: no change\n",
    "filtered_by_quality = utils.filter_quality(filtered_by_missingness)\n",
    "\n",
    "records = filtered_by_quality\n",
    "\n",
    "# Getting the number of samples and SNPs\n",
    "num_samples = len(vcf_file.header.samples)\n",
    "num_snps = len(records)\n",
    "\n",
    "# Create an empty genotype matrix\n",
    "G = np.empty((num_samples, num_snps), dtype=int)\n",
    "\n",
    "# Fill the matrix\n",
    "for j, record in enumerate(records):\n",
    "    for i, sample in enumerate(vcf_file.header.samples):\n",
    "        genotype = record.samples[sample].allele_indices\n",
    "        # Biallelic SNPs\n",
    "        if genotype == (0, 0):\n",
    "            G[i][j] = 0\n",
    "        elif genotype in [(0, 1), (1, 0)]:\n",
    "            G[i][j] = 1\n",
    "        else:\n",
    "            G[i][j] = 2\n",
    "\n",
    "\n",
    "U, S, Vt = np.linalg.svd(G, full_matrices = False)\n",
    "\n",
    "np.savez_compressed(\"../processed_data/real/ALL.chr21.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes)_SVD.npz\", U = U, S = S, Vt = Vt)\n",
    "np.savez_compressed(\"../processed_data/real/ALL.chr21.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.npz\", G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We simulate data of various sizes, sequence lengths, and sparsities\n",
    "sample_sizes = [50, 200, 1000, 10000]\n",
    "sequence_lengths = [10**x for x in range(4, 8)]\n",
    "mutation_rates = [1e-4, 1e-6, 1e-8] #lower mutation rate = sparser\n",
    "recombination_rate = 1e-7\n",
    "\n",
    "G_simulated = {}\n",
    "\n",
    "for n in sample_sizes:\n",
    "    for length in sequence_lengths:\n",
    "        for mut_rate in mutation_rates:\n",
    "            G = utils.simulate_diploid_genotypes(n, length, mut_rate, recombination_rate)\n",
    "            if G.shape[1] >= 10:\n",
    "                G_simulated[(n, length, mut_rate)] = G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save simulated data\n",
    "data_dir = \"../processed_data/simulated\"\n",
    "for (params, G) in G_simulated.items():\n",
    "    file_name = f\"{params[0]}_{params[1]}_{params[2]}.genotypes.npz\"\n",
    "    np.savez_compressed(f\"{data_dir}/{file_name}\", G)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Summary Statistics "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Real data from 1000 genomes project\n",
    "G = np.load(\"../processed_data/real/ALL.chr21.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.npz\")['arr_0']\n",
    "svd_data = np.load(\"../processed_data/real/ALL.chr21.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes)_SVD.npz\")\n",
    "\n",
    "U, S, Vt = svd_data[\"U\"], svd_data[\"S\"], svd_data[\"Vt\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.5923702615823612, 0.2359966071586969, 0.17163313125894197)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Calculate sparsity statistics for G\n",
    "utils.count_genotypes(G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load simulated data\n",
    "\n",
    "sample_sizes = [50, 200, 1000, 10000]\n",
    "sequence_lengths = [10**x for x in range(4, 8)] \n",
    "mutation_rates = [1e-4, 1e-6, 1e-8] #lower mutation rate = sparser\n",
    "recombination_rate = 1e-7\n",
    "\n",
    "G_simulated = {}\n",
    "\n",
    "simulated_dir = \"../processed_data/simulated/\"\n",
    "for n in sample_sizes:\n",
    "    for length in sequence_lengths:\n",
    "        for mut_rate in mutation_rates:\n",
    "            file_name = f\"{n}_{length}_{mut_rate}.genotypes.npz\"\n",
    "            file_dir = simulated_dir + file_name \n",
    "            try:\n",
    "                G = np.load(file_dir)[\"arr_0\"]\n",
    "                G_simulated[(n, length, mut_rate)] = G\n",
    "            except:\n",
    "                pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate sparsity statistics\n",
    "G_simulated_props = {key: utils.count_genotypes(G_simulated[key]) for key in G_simulated.keys()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Parameters</th>\n",
       "      <th>Shape of G</th>\n",
       "      <th>Proportion Zeros</th>\n",
       "      <th>Proportion Ones</th>\n",
       "      <th>Proportion Twos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(50, 10000, 0.0001)</td>\n",
       "      <td>(50, 15)</td>\n",
       "      <td>0.537333</td>\n",
       "      <td>0.156000</td>\n",
       "      <td>0.306667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(50, 100000, 0.0001)</td>\n",
       "      <td>(50, 204)</td>\n",
       "      <td>0.784216</td>\n",
       "      <td>0.154314</td>\n",
       "      <td>0.061471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(50, 1000000, 0.0001)</td>\n",
       "      <td>(50, 1889)</td>\n",
       "      <td>0.761207</td>\n",
       "      <td>0.189137</td>\n",
       "      <td>0.049656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(50, 1000000, 1e-06)</td>\n",
       "      <td>(50, 35)</td>\n",
       "      <td>0.668571</td>\n",
       "      <td>0.224571</td>\n",
       "      <td>0.106857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(50, 10000000, 0.0001)</td>\n",
       "      <td>(50, 17331)</td>\n",
       "      <td>0.786631</td>\n",
       "      <td>0.176902</td>\n",
       "      <td>0.036468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>(50, 10000000, 1e-06)</td>\n",
       "      <td>(50, 232)</td>\n",
       "      <td>0.752931</td>\n",
       "      <td>0.181121</td>\n",
       "      <td>0.065948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>(200, 10000, 0.0001)</td>\n",
       "      <td>(200, 22)</td>\n",
       "      <td>0.693864</td>\n",
       "      <td>0.147500</td>\n",
       "      <td>0.158636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>(200, 100000, 0.0001)</td>\n",
       "      <td>(200, 178)</td>\n",
       "      <td>0.901348</td>\n",
       "      <td>0.088399</td>\n",
       "      <td>0.010253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>(200, 1000000, 0.0001)</td>\n",
       "      <td>(200, 2049)</td>\n",
       "      <td>0.853675</td>\n",
       "      <td>0.114422</td>\n",
       "      <td>0.031903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>(200, 1000000, 1e-06)</td>\n",
       "      <td>(200, 20)</td>\n",
       "      <td>0.627750</td>\n",
       "      <td>0.295750</td>\n",
       "      <td>0.076500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>(200, 10000000, 0.0001)</td>\n",
       "      <td>(200, 26446)</td>\n",
       "      <td>0.784120</td>\n",
       "      <td>0.148175</td>\n",
       "      <td>0.067705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>(200, 10000000, 1e-06)</td>\n",
       "      <td>(200, 214)</td>\n",
       "      <td>0.807734</td>\n",
       "      <td>0.130584</td>\n",
       "      <td>0.061682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>(1000, 10000, 0.0001)</td>\n",
       "      <td>(1000, 33)</td>\n",
       "      <td>0.862848</td>\n",
       "      <td>0.111091</td>\n",
       "      <td>0.026061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>(1000, 100000, 0.0001)</td>\n",
       "      <td>(1000, 279)</td>\n",
       "      <td>0.861749</td>\n",
       "      <td>0.079269</td>\n",
       "      <td>0.058982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>(1000, 1000000, 0.0001)</td>\n",
       "      <td>(1000, 2557)</td>\n",
       "      <td>0.899564</td>\n",
       "      <td>0.081402</td>\n",
       "      <td>0.019034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>(1000, 1000000, 1e-06)</td>\n",
       "      <td>(1000, 18)</td>\n",
       "      <td>0.889167</td>\n",
       "      <td>0.047611</td>\n",
       "      <td>0.063222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>(1000, 10000000, 0.0001)</td>\n",
       "      <td>(1000, 34599)</td>\n",
       "      <td>0.823802</td>\n",
       "      <td>0.137456</td>\n",
       "      <td>0.038742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>(1000, 10000000, 1e-06)</td>\n",
       "      <td>(1000, 372)</td>\n",
       "      <td>0.805820</td>\n",
       "      <td>0.127419</td>\n",
       "      <td>0.066761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>(10000, 10000, 0.0001)</td>\n",
       "      <td>(10000, 43)</td>\n",
       "      <td>0.765809</td>\n",
       "      <td>0.130233</td>\n",
       "      <td>0.103958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>(10000, 100000, 0.0001)</td>\n",
       "      <td>(10000, 399)</td>\n",
       "      <td>0.834666</td>\n",
       "      <td>0.112971</td>\n",
       "      <td>0.052363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>(10000, 1000000, 0.0001)</td>\n",
       "      <td>(10000, 3618)</td>\n",
       "      <td>0.891057</td>\n",
       "      <td>0.074841</td>\n",
       "      <td>0.034102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>(10000, 1000000, 1e-06)</td>\n",
       "      <td>(10000, 40)</td>\n",
       "      <td>0.832770</td>\n",
       "      <td>0.123895</td>\n",
       "      <td>0.043335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>(10000, 10000000, 0.0001)</td>\n",
       "      <td>(10000, 41561)</td>\n",
       "      <td>0.883166</td>\n",
       "      <td>0.080379</td>\n",
       "      <td>0.036455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>(10000, 10000000, 1e-06)</td>\n",
       "      <td>(10000, 432)</td>\n",
       "      <td>0.868766</td>\n",
       "      <td>0.090207</td>\n",
       "      <td>0.041027</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Parameters      Shape of G  Proportion Zeros  \\\n",
       "0         (50, 10000, 0.0001)        (50, 15)          0.537333   \n",
       "1        (50, 100000, 0.0001)       (50, 204)          0.784216   \n",
       "2       (50, 1000000, 0.0001)      (50, 1889)          0.761207   \n",
       "3        (50, 1000000, 1e-06)        (50, 35)          0.668571   \n",
       "4      (50, 10000000, 0.0001)     (50, 17331)          0.786631   \n",
       "5       (50, 10000000, 1e-06)       (50, 232)          0.752931   \n",
       "6        (200, 10000, 0.0001)       (200, 22)          0.693864   \n",
       "7       (200, 100000, 0.0001)      (200, 178)          0.901348   \n",
       "8      (200, 1000000, 0.0001)     (200, 2049)          0.853675   \n",
       "9       (200, 1000000, 1e-06)       (200, 20)          0.627750   \n",
       "10    (200, 10000000, 0.0001)    (200, 26446)          0.784120   \n",
       "11     (200, 10000000, 1e-06)      (200, 214)          0.807734   \n",
       "12      (1000, 10000, 0.0001)      (1000, 33)          0.862848   \n",
       "13     (1000, 100000, 0.0001)     (1000, 279)          0.861749   \n",
       "14    (1000, 1000000, 0.0001)    (1000, 2557)          0.899564   \n",
       "15     (1000, 1000000, 1e-06)      (1000, 18)          0.889167   \n",
       "16   (1000, 10000000, 0.0001)   (1000, 34599)          0.823802   \n",
       "17    (1000, 10000000, 1e-06)     (1000, 372)          0.805820   \n",
       "18     (10000, 10000, 0.0001)     (10000, 43)          0.765809   \n",
       "19    (10000, 100000, 0.0001)    (10000, 399)          0.834666   \n",
       "20   (10000, 1000000, 0.0001)   (10000, 3618)          0.891057   \n",
       "21    (10000, 1000000, 1e-06)     (10000, 40)          0.832770   \n",
       "22  (10000, 10000000, 0.0001)  (10000, 41561)          0.883166   \n",
       "23   (10000, 10000000, 1e-06)    (10000, 432)          0.868766   \n",
       "\n",
       "    Proportion Ones  Proportion Twos  \n",
       "0          0.156000         0.306667  \n",
       "1          0.154314         0.061471  \n",
       "2          0.189137         0.049656  \n",
       "3          0.224571         0.106857  \n",
       "4          0.176902         0.036468  \n",
       "5          0.181121         0.065948  \n",
       "6          0.147500         0.158636  \n",
       "7          0.088399         0.010253  \n",
       "8          0.114422         0.031903  \n",
       "9          0.295750         0.076500  \n",
       "10         0.148175         0.067705  \n",
       "11         0.130584         0.061682  \n",
       "12         0.111091         0.026061  \n",
       "13         0.079269         0.058982  \n",
       "14         0.081402         0.019034  \n",
       "15         0.047611         0.063222  \n",
       "16         0.137456         0.038742  \n",
       "17         0.127419         0.066761  \n",
       "18         0.130233         0.103958  \n",
       "19         0.112971         0.052363  \n",
       "20         0.074841         0.034102  \n",
       "21         0.123895         0.043335  \n",
       "22         0.080379         0.036455  \n",
       "23         0.090207         0.041027  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate dataframe\n",
    "df_keys = [{'Parameters': key, 'Shape of G': G_simulated[key].shape,\n",
    "         'Proportion Zeros': value[0], 'Proportion Ones': value[1], 'Proportion Twos': value[2]}\n",
    "        for key, value in G_simulated_props.items()]\n",
    "\n",
    "G_df = pd.DataFrame(df_keys)\n",
    "\n",
    "# Write to LaTeX\n",
    "with open('../figures/sim_data_table.txt', 'w') as f:\n",
    "    f.write(G_df.to_latex(index=False,\n",
    "                          bold_rows=False,\n",
    "                          float_format=\"%.2f\"))\n",
    "\n",
    "G_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Algorithms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Darnell et al: ARSVD\n",
    "\n",
    "Paper: https://www.jmlr.org/papers/volume18/15-143/15-143.pdf \n",
    "\n",
    "Implementation: https://github.com/gdarnell/arsvd/blob/master/dimension_reduction.py \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rsvd(X, dstar, power_iters=2, delta=10):\n",
    "\t\"\"\" Perform rsvd algorithm on input matrix.\n",
    "\t\tMethod must be supplied dstar.\n",
    "\t\tReturns truncated svd (U,S,V).\n",
    "\tParameters\n",
    "\t----------\n",
    "\tX : int matrix\n",
    "    \tMatrix of n x m integers, where m <= n. If n < m,\n",
    "    \tmatrix will be transposed to enforce m <= n.\n",
    "   \tdstar : int\n",
    "   \t\tThe latent (underlying) matrix rank that will be\n",
    "   \t\tused to truncate the larger dimension (m).\n",
    "   \tpower_iters : int\n",
    "   \t\tdefault: 2\n",
    "   \t\tNumber of power iterations used (random matrix multiplications)\n",
    "   \tdelta : int\n",
    "   \t\tdefault: 10\n",
    "   \t\toversampling parameter (to improve numerical stability)\n",
    "    Returns\n",
    "\t-------\n",
    "\tint matrix\n",
    "    \tMatrix of left singular vectors.\n",
    "    int matrix\n",
    "    \tMatrix of singular values.\n",
    "    int matrix\n",
    "    \tMatrix of right singular vectors.\n",
    "    \"\"\"\n",
    "\ttranspose = False \n",
    "\tif X.shape[0] < X.shape[1]:\n",
    "\t\tX = X.T \n",
    "\t\ttranspose = True \n",
    "\n",
    "\tif power_iters < 1:\n",
    "\t\tpower_iters = 1\n",
    "\n",
    "\t# follows manuscript notation as closely as possible\n",
    "\tP = np.random.randn(X.shape[0],dstar+delta)\t\n",
    "\tfor i in range(power_iters):\n",
    "\t\tP = np.dot(X.T,P)\n",
    "\t\tP = np.dot(X,P)\n",
    "\tQ,R = np.linalg.qr(P)\n",
    "\tB = np.dot(Q.T,X)\n",
    "\tU,S,V = linalg.svd(B)\n",
    "\tU = np.dot(Q, U)\n",
    "\n",
    "\t# Remove extra dimensionality incurred by delta\n",
    "\tU = U[:, 0:dstar]\n",
    "\tS = S[0:dstar]\n",
    "\n",
    "\treturn V.T, S, U.T if transpose else U, S, V\n",
    "\n",
    "\n",
    "def stabilityMeasure(X, d_max, B=5, power_iters=2):\n",
    "\t\"\"\" Calculate stability of \n",
    "\tParameters\n",
    "\t----------\n",
    "\tX : int matrix\n",
    "\t\tinput matrix to determine rank of\n",
    "\td_max : int\n",
    "\t\tupper bound rank to estimate\n",
    "\tB : int\n",
    "\t\tdefault: 5\n",
    "\t\tnumber of projections to correlate\n",
    "\tpower_iters : int\n",
    "\t\tdefault: 2\n",
    "   \t\tNumber of power iterations used (random matrix multiplications)\n",
    "\tReturns\n",
    "\t-------\n",
    "\tint\n",
    "\t\tLatent (lower-dimensional) matrix rank\n",
    "\t\"\"\"\n",
    "\tsingular_basis = np.zeros((B,X.shape[0],d_max))\n",
    "\t# calculate singular basis under multiple projections\n",
    "\tfor i in range(B):\n",
    "\t\tU = rsvd(X,d_max)[0]\n",
    "\t\tsingular_basis[i,:,:] = U[:,0:d_max]\n",
    "\n",
    "\t# calculate score for each singular vector\n",
    "\tstability_vec = np.zeros((d_max))\n",
    "\tfor k in range(d_max):\n",
    "\t\tstability = 0\n",
    "\t\tfor i in range(0,B-1):\n",
    "\t\t\tfor j in range(i+1,B):\n",
    "\t\t\t\tcorr = stats.spearmanr(singular_basis[i,:,k],singular_basis[j,:,k])[0]\n",
    "\t\t\t\tstability = stability + abs(corr)\n",
    "\t\tN = B*(B-1)/2\n",
    "\t\tstability = stability/N\n",
    "\t\tstability_vec[k] = stability\n",
    "\n",
    "\t# wilcoxon rank-sum test p-values\n",
    "\tp_vals = np.zeros(d_max-2)\n",
    "\tfor k in range(2,d_max):\n",
    "\t\tp_vals[k-2] = stats.ranksums(stability_vec[0:k-1],stability_vec[k-1:d_max])[1]\n",
    "\n",
    "\tdstar = np.argmin(p_vals)\n",
    "\t\n",
    "\treturn dstar"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saibaba et al: GSVD\n",
    "\n",
    "Paper: https://www.osti.gov/servlets/purl/1769894 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gsvd():\n",
    "    pass "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Benchmarking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def benchmark(f: Callable, k: int, data: dict, *args):\n",
    "    \"\"\" \n",
    "    Benchmarks performance of an algorithm on the given datasets.\n",
    "\n",
    "    Inputs:\n",
    "    - f (Callable): the PCA algorithm\n",
    "        - should take in an argument for data and an argument for number of principal components\n",
    "    - k (int): the desired number of principal components\n",
    "    - data (dict): the datasets, where data[params] = dataset\n",
    "\n",
    "    Returns:\n",
    "    - dict of accuracy metrics by dataset\n",
    "    \"\"\"\n",
    "    accuracy_metrics = {}\n",
    "\n",
    "    for (params, dataset) in data.items():\n",
    "        # Calculate proportion of total variance for top k PCs\n",
    "        accuracy_key = (params, k)\n",
    "        pcs = f(dataset, k, *args)\n",
    "        accuracy_metrics[accuracy_key] = utils.compute_prop_variance(pcs, dataset)\n",
    "\n",
    "    return accuracy_metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Values for k\n",
    "k_vals = [1, 2, 3, 5, 10, 20, 30]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ARSVD\n",
    "def pca_arsvd(X, dstar, k):\n",
    "    # Need to check for transpose or not\n",
    "    U, S, VT = rsvd(X, dstar)\n",
    "\n",
    "    # Compute top k principal components\n",
    "    \n",
    "\n",
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
